File: sample-ios/sample-ios/AnalyticsProvider.swift
import analytics



class AnalyticsProvider {
    
    static var analytics: Analytics = {
        return Analytics.Builder()
            .addConsumer(consumer: FooConsumer())
            .addConsumerInterceptor(consumerInterceptor: FooInterceptor())
            .addInterceptor(interceptor: CommonInterceptor())
            .setExceptionHandler(exceptionHandler: { throwable in print("error" + throwable.message!) })
            .setDebugLog(shouldEnable: true)
            .build()
    }()
}

class CommonInterceptor: Interceptor{
    
    func intercept(event: Event, completionHandler: @escaping (Event?, Error?) -> Void) {
        
        let outParams = (event.params as NSDictionary).mutableCopy() as! NSMutableDictionary
        outParams.addEntries(from: ["deviceId":"ID", "userId": "ID"])
        return completionHandler(Event(event: event.event, sum: event.sum,
                                       params: outParams as! [String: Any]), nil);
    }
}

class FooInterceptor: ConsumerInterceptor{
    func intercept(consumer: AnalyticsConsumer, event: Event, completionHandler: @escaping (Event?, Error?) -> Void) {
        
        let outParams = (event.params as NSDictionary).mutableCopy() as! NSMutableDictionary
        if(consumer is FooConsumer){
            outParams.addEntries(from: ["fooUserId":"ID"])
        }
        return completionHandler( Event(event: event.event, sum: event.sum,
                                        params: outParams as! [String: Any]) , nil);
    }
}


class FooConsumer : AnalyticsConsumer {
    func canAccept(event: String) -> Bool {
        return true
    }
    
    func acceptEvent(event: AnalyticsEvent) {
        print("🚀 " + String(describing: type(of: self)) + " event accepted")
    }
}

File: sample-android/src/main/java/com/sample/analytics/analytics/AnalyticsProvider.kt
package com.sample.analytics.analytics

import analytics.Analytics
import analytics.AnalyticsConsumer
import analytics.BaseAnalyticsConsumer
import analytics.ConsumerInterceptor
import analytics.Events.ALL
import analytics.Interceptor
import analytics.consumer.firebase.FirebaseConsumer
import analytics.consumer.yandex.YandexMetricaConsumer
import analytics.event.Event
import com.sample.analytics.analytics.event.eventArray
import com.google.firebase.analytics.FirebaseAnalytics
import com.sample.analytics.App
import com.sample.analytics.analytics.event.CheckoutAnalytics
import java.util.logging.Logger

object AnalyticsProvider {

  private lateinit var INSTANCE: Analytics

  private val YANDEX_EVENTS = listOf(*eventArray<CheckoutAnalytics.Events>())
  private val FIREBASE_EVENTS = listOf(CheckoutAnalytics.Events.ADDRESS_SELECTED.event)

  fun provide(): Analytics =
    if (::INSTANCE.isInitialized) INSTANCE
    else {

      INSTANCE = Analytics.Builder()
          .addConsumer(FooConsumer())
          .addConsumer(YandexMetricaConsumer(YANDEX_EVENTS))
          .addConsumer(FirebaseConsumer(FirebaseAnalytics.getInstance(App.INSTANCE), FIREBASE_EVENTS))
          .addInterceptor(CommonInterceptor())
          .addConsumerInterceptor(FooConsumerInterceptor())
//          .setExceptionHandler { error -> logger.warning(error.message ?: "") }
          .setDebugLog(true)
          .build()
      INSTANCE
    }

}

class CommonInterceptor : Interceptor {
  override suspend fun intercept(event: Event) =
    event.copy(params = event.params + mapOf("deviceId" to "ID", "userId" to "ID"))
}

//private val logger = Logger.getLogger("logger_analytics_sample")

class FooConsumer : BaseAnalyticsConsumer(ALL) {
  override fun acceptDefaultEvent(event: Event) {
//    logger.info("${this::class.simpleName} acceptEvent: $event")
  }
}

class FooConsumerInterceptor : ConsumerInterceptor {
  override suspend fun intercept(
    consumer: AnalyticsConsumer,
    event: Event
  ) = takeIf { consumer is FooConsumer }
      ?.let { event.copy(params = event.params + ("fooUserId" to "ID")) }
      ?: event
}

